tcl/org.testeditor.tcl.dsl/src/org/testeditor/tcl/dsl/validation/TclValidator.xtend

```diff
@@ -361,15 +384,16 @@ class TclValidator extends AbstractTclValidator {
 	@Check
 	def void checkVariableReferenceUsage(TclModel tclModel) {
 		val stringTypeReference = typeReferences.getTypeForName(String, tclModel)
-		val environmentParams = tclModel.envParams.map[name].toSet
-		val actualTypeMap = newHashMap
-		environmentParams.forEach [
-			actualTypeMap.put(it, #{stringTypeReference})
+		val availableParamNames = tclModel.envParams.map[name].toSet
+		val availableParamNamesTypeMap = newHashMap
+		availableParamNames.forEach [
+			availableParamNamesTypeMap.put(it, #{stringTypeReference})
 		]
 		tclModel.test.steps.map[contexts].flatten.forEach [
-			checkAllVariableReferencesAreKnownParameters(environmentParams,
-				"Dereferenced variable must be a required environment variable")
-			checkAllVariableReferencesOnTypeEquality(actualTypeMap)
+			checkAllVariableReferencesAreKnownParameters(availableParamNamesTypeMap.keySet,
+				"Dereferenced variable must be a required environment variable or a previously assigned variable")
+			availableParamNamesTypeMap.putAll(collectAssignmentVariables)
```

Comment:
This is pretty hard to understand. You reuse the `availableParamNamesTypeMap` to put additional values in it for the next check. This should be refactored or at least commented.